### ISSUE TITLE
Proposal: Implement heroku cache using objectstore

### ISSUE THREAD

**Issue by Joshua-Anderson** [deis/slugbuilder#95]
Heroku offers a $CACHE_DIR for buildpacks to store files that will be persisted across builds ([reference](https://devcenter.heroku.com/articles/buildpack-api#caching)).

Builder offers the directory for buildpack compatibility, but the contents of the files are not persisted when the slug builder exits.

Deis V1 used the same builder for all apps and stored the cache directory in a volume, so unless it got rescheduled it cached properly.

For deis workflow however, I recommend uploading the cache dir after a build to the builder object store bucket and downloading it at the start of the next build.

Sometimes, the cache becomes corrupted and needs to be deleted. Sometimes users do not want to cache the cache directory. In that case, they could set the `DEIS_BUILDPACK_CACHE` environmental variable to `false`, which would trigger the slugbuilder to delete the cache if it does exist and skip uploading the cache at the end of the build.

**Comment by kmala** [comment]
i had thought about this approach before but here we would need to upload the cache after a build and download the cache before a build which would mean extra work than if we are directly downloading dependencies or building  without using the cache.

**Comment by Joshua-Anderson** [comment]
There are some situation when using the cache would be slower, however, in most situations, it is significantly faster.
1. You can store stuff in the same datacenter as your cluster. For example, you can store dependencies in GCS when using GKE, which is likely faster than anywhere else.
2. Many languages have post download steps. Most ruby, python, and nodejs projects have C++ dependencies that compile on installation and take ages doing so. In these cases, pulling from the cache saves literal minutes.
3. We can be smart about uploading dependencies, Travis-CI's caching tool is open source and we can take inspiration from it ([casher](https://github.com/travis-ci/casher)). What they do is check the md5 of all the files in the cache to detect changes before uploading. In addition, we can have the builder launch the app after it finishes building and have the slugbuilder run in the background uploading.

**Comment by kmala** [comment]
yes there are definitely few situations where it is better than not having cache at all.It's just we need to think about all the pros and cons before starting to work on this.

**Comment by Joshua-Anderson** [comment]
I can do some performance testing if you would like. Anecdotally from my travis-ci experience I've seen noticeable improvements with nodejs, go, python, and ruby when caching on travis-ci.

**Comment by nathansamson** [comment]
In rails apps a common step is compiling assets. In my case this takes 700-800s on deis V1 after a reschedule of builder. (lots of javascript libraries; I probably should take some time to optimize this but thats another story)
Downloading from S3 would be significantly faster as generally a second build only takes around 10s if nothing has changed (so even if downloading the cache takes a minute I gain a lot by this)

Having a per app setting (or action) to (temporirly) disable the cache would probably be a good idea. Apparantly Heroku has something like that as a plugin only... (https://github.com/heroku/heroku-repo)

**Comment by jeroenvisser101** [comment]
@kmala @Joshua-Anderson in Rails, this makes a very large difference, although I can't speak for many others, downloading dependencies happens almost everywhere, and that would benefit greatly from caching. Not just assets as @nathansamson mentioned, but also just downloading gems. Deploying our current project to Heroku takes about 15-20 seconds, while the same deploy can take up to 15 minutes on Deis.

As for the implementation, storing them could be in GCS, but even when storing them inside the cluster (with the risk of losing when the pod restarts) would be better than what we have right now. [Flynn does this](https://github.com/flynn/flynn/tree/master/blobstore) too, and they just [upload](https://github.com/flynn/flynn/blob/c10e46113b92abc2bc7565699cc0e4892b5f69bb/slugbuilder/builder/build.sh#L244-L255)/[download](https://github.com/flynn/flynn/blob/c10e46113b92abc2bc7565699cc0e4892b5f69bb/slugbuilder/builder/build.sh#L106-L108) the files to/from that blobstore. We could do something similar (whether using S3/GCS or Minio), and gain a lot out of this. Most buildpacks ([ruby](https://github.com/heroku/heroku-buildpack-ruby/blob/master/lib/language_pack/ruby.rb#L840-L933), [php](https://github.com/heroku/heroku-buildpack-php/blob/master/bin/compile#L72), [nodejs](https://github.com/heroku/heroku-buildpack-nodejs/blob/master/bin/compile#L102-L117), [python](https://github.com/heroku/heroku-buildpack-python/blob/master/bin/compile#L167-L177), [go (multiple places)](https://github.com/heroku/heroku-buildpack-go/blob/master/bin/compile#L317-L348)) explicitly use this cache to store artifacts that can be reused, but since we don't persist this between deploys, we lose the performance that it was designed to gain.

There's also other caches ([although currently disabled](https://github.com/heroku/heroku-buildpack-ruby/blob/master/lib/language_pack/ruby.rb#L507)) that we could implement later, but those would only benefit new apps.
